### DISCUSSION TITLE
Context Window Size and RAG Improvement Ideas

### DISCUSSION THREAD

**Discussion by tomazfernandes-up** [open-webui/open-webui#4275] [Ideas]
Hi, first of all, congratulations on the great work and amazing project!

I'm working on a project to enable using `OpenWebUI's RAG` with local `Ollama` and `Llama 3.1` 8B to ask questions about a local `Git` repository and documentation so it can give relevant answers based on the documents.

I've looked around and it seems there are quite a few improvements being considered, so please forgive me if some of those are already being solved. 

Please let me know if these suggestions make sense, I'd be happy to open a few issues and contribute PRs if so.

Thanks!

- **Allow changing the context window size in the UI**
I was able to change the context window by manipulating the payload, and it works ok with 16k context window on a 16MB Mac. Would make sense to include this as an option in the Models tab of the Admin Panel?

-  **Allow including or excluding file types in RAG scan**
In my use case it's helpful to for example exclude test files from the scan, since they tend to be larger files that potentially fill up the context. An option in the Documents menu might help filter out irrelevant files and improve retrieval relevance.

- **Use the chat history when querying the vector database so we can ask follow-up questions with documents in the context**
The current implementation only takes the last prompt into consideration when querying the vector database. This makes it not possible to ask follow-up questions about the documents, since the question might not contain the relevant context for selecting the documents to be retrieved.

- **Store all documents in the same collection so we don't need to retrieve all of them and iterate to process relevance**
The current solution retrieves and iterates on all documents to classify relevance, which is not performant. While I do understand that this is possibly to support the user to chose only a handful of documents for a query, it might be interesting to have an `all documents` collection that would rely on the vector DB mechanism for this retrieval, or some other similar solution.

- **Show the file name and ideally the entire file in the answer sources**
Since LLMs should be fact-checked for any serious use, it might be nice to allow showing the document names and the full documents in the answer sources when using all documents. Maybe it'd also be possible to do some code highlighting and formatting, though that might be more complicated to implement.

**Comment by Pranaviee** [comment]
Hi,sir
This is similar to my work. I made an agentic rag framework for code generation. The knowledge base is github repo.
But the final llm which is taking all this contexts to answer the query is not at all reasoning the answer seems vague and llama3 is not compleetely looking the prompt.

is this due to long contexts?
